Fix udev firmware events dependencies

The "firmware" udev builtin command was dropped from the systemd by the commit:
https://github.com/systemd/systemd/commit/be2ea723b1d023b3d385d3b791ee4607cbfb20ca

But because of that the firmware loading events can be triggered too late,
waiting for other events, e.g. module loading, to be processed first. But the
device driver can in turn wait for the firmware to be loaded. So a timeout could
happen and the device initialization will be aborted.

Therefore let's bring back the code that skips dependency on the parent devices
for firmware loading events.

See: https://github.com/systemd/systemd/issues/18259

diff --git a/src/udev/udevd.c b/src/udev/udevd.c
index 9320284be6..b618fd5e1f 100644
--- a/src/udev/udevd.c
+++ b/src/udev/udevd.c
@@ -127,6 +127,8 @@ typedef struct Event {
         uint64_t seqnum;
         uint64_t blocker_seqnum;
 
+        bool nodelay;
+
         sd_event_source *timeout_warning_event;
         sd_event_source *timeout_event;
 
@@ -789,6 +791,10 @@ static int event_is_blocked(Event *event) {
         assert(event->manager);
         assert(event->blocker_seqnum <= event->seqnum);
 
+        /* allow to bypass the dependency tracking */
+        if (event->nodelay)
+                return false;
+
         if (event->blocker_seqnum == event->seqnum)
                 /* we have checked previously and no blocker found */
                 return false;
@@ -979,6 +985,7 @@ static int event_queue_insert(Manager *manager, sd_device *dev) {
         _cleanup_(sd_device_unrefp) sd_device *clone = NULL;
         Event *event;
         uint64_t seqnum;
+        const char *val;
         int r;
 
         assert(manager);
@@ -1013,6 +1020,13 @@ static int event_queue_insert(Manager *manager, sd_device *dev) {
                 .state = EVENT_QUEUED,
         };
 
+        r = sd_device_get_subsystem(dev, &val);
+        if (r < 0)
+                return r;
+
+        if (streq(val, "firmware"))
+                event->nodelay = true;
+
         if (LIST_IS_EMPTY(manager->events)) {
                 r = touch("/run/udev/queue");
                 if (r < 0)
